lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -265,6 +273,142 @@ getBlocks logTrace logic recoveryHeadersMessage enqueue nodeId tipHeaderHash che
+    writeStreamEnd blockChan = atomically $ Conc.writeTBQueue blockChan StreamEnd
+
+    mkStreamStart :: [HeaderHash] -> HeaderHash -> MsgStream
+    mkStreamStart chain wantedBlock =
+        MsgStart $ MsgStreamStart
+        { mssFrom = chain
+        , mssTo = wantedBlock
+        , mssWindow = streamWindow
+        }
+
+    requestBlocks :: Conc.TVar Bool -> Conc.TBQueue StreamEntry -> IO (Conc.TVar (OQ.PacketStatus ()))
+    requestBlocks fallBack blockChan = do
+        convMap <- enqueue (MsgRequestBlocks (S.singleton nodeId))
+                            (\_ _ -> (Conversation $ \it -> requestBlocksConversation blockChan it `finally` writeStreamEnd blockChan) :|
+                                      [(Conversation $ \it -> requestBatch fallBack blockChan it `finally` writeStreamEnd blockChan)]
+                                     )
+        case M.lookup nodeId convMap of
+            Just tvar -> pure tvar
+            -- FIXME shouldn't have to deal with this.
+            -- One possible solution: do the block request in response to an
+            -- unsolicited header, so that's it's all done in one conversation,
+            -- and so there's no need to even track the 'nodeId'.
+            Nothing   -> throwM $ DialogUnexpected $ "requestBlocks did not contact given peer"
+
+    requestBatch
+        :: Conc.TVar Bool
+        -> Conc.TBQueue StreamEntry
+        -> ConversationActions MsgGetBlocks MsgBlock
+        -> IO ()
+    requestBatch fallBack blockChan _ = do
+        -- The peer doesn't support streaming, we need to fall back to batching but
+        -- the current conversation is unusable since there is no way for us to learn
+        -- which blocks we shall fetch.
+        atomically $ writeTVar fallBack True
+        atomically $ Conc.writeTBQueue blockChan StreamEnd
+        return ()
+
+    requestBlocksConversation
+        :: Conc.TBQueue StreamEntry
+        -> ConversationActions MsgStream MsgStreamBlock
+        -> IO ()
+    requestBlocksConversation blockChan conv = do
+        let newestHash = headerHash tipHeader
+        traceWith logTrace (Debug,
+            sformat ("streamBlocks: Requesting stream of blocks from "%listJson%" to "%shortHashF)
+                     checkpoints
+                     newestHash)
+        send conv $ mkStreamStart checkpoints newestHash
+        bvd <- Logic.getAdoptedBVData logic
+        retrieveBlocks bvd blockChan conv streamWindow
+        return ()
+
+    halfStreamWindow = max 1 $ streamWindow `div` 2
+
+    -- A piece of the block retrieval conversation in which the blocks are
+    -- pulled in one-by-one.
+    retrieveBlocks
+        :: BlockVersionData
+        -> Conc.TBQueue StreamEntry
+        -> ConversationActions MsgStream MsgStreamBlock
+        -> Word32
+        -> IO ()
+    retrieveBlocks bvd blockChan conv window = do
+        window' <- if window < halfStreamWindow
+                          then do
```

Comment:
That's being picky ;), but the indentation is off here.

Reply:
Will fix.